ui/ui-frontend/projects/vitamui-library/src/app/modules/services/access-contract.service.ts

```diff
@@ -36,26 +36,69 @@
  */
 import { HttpHeaders, HttpParams, HttpResponse } from '@angular/common/http';
 import { Injectable } from '@angular/core';
-import { Observable, Subject } from 'rxjs';
-import { tap } from 'rxjs/operators';
-import { AccessContract, SearchService, VitamUISnackBarService } from 'vitamui-library';
-import { AccessContractApiService } from '../core/api/access-contract-api.service';
+import { BehaviorSubject, Observable, of, pipe, Subject, switchMap, UnaryFunction } from 'rxjs';
+import { map, tap } from 'rxjs/operators';
+import { AccessContractApiService } from '../api/access-contract-api.service';
+import { AccessContract } from '../../../lib/models/access-contract.interface';
+import { SearchService } from '../vitamui-table';
+import { VitamUISnackBarService } from '../components/vitamui-snack-bar';
+import { ExternalParametersService } from '../externalParameters.service';
+import { ExternalParameters } from '../externalParameters.enum';
 
 @Injectable({
   providedIn: 'root',
 })
 export class AccessContractService extends SearchService<AccessContract> {
+  private _currentAccessContractId: BehaviorSubject<string> = new BehaviorSubject(null);
+  private _currentAccessContract: BehaviorSubject<AccessContract> = new BehaviorSubject(null);
+
+  private loadAccessContractIdIfAbsent: UnaryFunction<Observable<string>, Observable<string>> = pipe(
+    switchMap((accessContractId) => {
+      if (accessContractId) return of(accessContractId);
+      return this.externalParameterService.getUserExternalParameters().pipe(
+        map((parameters) => parameters.get(ExternalParameters.PARAM_ACCESS_CONTRACT)),
+        tap((accessContractId) => this.setCurrentAccessContractId(accessContractId)),
+      );
+    }),
+  );
+  private loadAccessContractIfAbsent: UnaryFunction<Observable<AccessContract>, Observable<AccessContract>> = pipe(
+    switchMap((accessContract) => {
+      if (accessContract) return of(accessContract);
+      return this.currentAccessContractId$.pipe(
+        switchMap((accessContractId) => this.get(accessContractId)),
+        tap((accessContract) => this.setCurrentAccessContract(accessContract)),
+      );
+    }),
+  );
+
+  /** Observable of current access contract ID */
+  currentAccessContractId$: Observable<string> = this._currentAccessContractId.asObservable().pipe(this.loadAccessContractIdIfAbsent);
+  /** Observable of current access contract */
+  currentAccessContract$: Observable<AccessContract> = this._currentAccessContract.asObservable().pipe(this.loadAccessContractIfAbsent);
+
   updated = new Subject<AccessContract>();
 
+  setCurrentAccessContractId(id: string) {
+    this._currentAccessContract.next(null);
+    this._currentAccessContractId.next(id);
+  }
+
+  private setCurrentAccessContract(accessContract: AccessContract) {
+    this._currentAccessContract.next(accessContract);
+    this._currentAccessContractId.next(accessContract.id);
+  }
```

Comment:
Il me semble que l'on peut simplifier par ça :
```suggestion
  /** Observable of current access contract ID */
  currentAccessContractId$: Observable<string> = this.externalParameterService.getUserExternalParameters().pipe(
    map((parameters) => parameters.get(ExternalParameters.PARAM_ACCESS_CONTRACT)),
    shareReplay(1),
  );
  /** Observable of current access contract */
  currentAccessContract$: Observable<AccessContract> = this.currentAccessContractId$.pipe(
    switchMap((accessContractId) => this.get(accessContractId)),
    shareReplay(1),
  );

  updated = new Subject<AccessContract>();
```

En supprimant `this.accessContractService.setCurrentAccessContractId(accessConctractId);` de archive.component.ts

Reply:
Ca ne va pas faire l'appel a chaque fois cette solution ?

Reply:
C'est MAJ